### PR TITLE
Dataset definition

### PR DESCRIPTION
**Pull request by alainamstutz**
Hi Zoe,
I implemented your initial suggestions (from the old repo). All worked fine - except when I tried to implement the metadatas.R structure, run via yaml, it bugs. I tried for a while, my setup looks like the same as yours (https://github.com/grpEHR/post-covid-events-ehrQL/blob/dataset-definition/analysis/metadates.R). 

Locally, I can run each command individually (opensafely exec r:latest analysis/metadates.R --output output/study_dates.json, then opensafely exec ehrql:v1 generate-dataset analysis/dataset_definition.py --output output/dataset.arrow) and that works fine, but when I run all together (opensafely run run_all ) it breaks (see error message), must have something to do with the setup/yaml actions?

Any idea?

### CODE REVIEW

**Comment by ZoeMZou** [comment]
> Hi Zoe, I implemented your initial suggestions (from the old repo). All worked fine - except when I tried to implement the metadatas.R structure, run via yaml, it bugs. I tried for a while, my setup looks like the same as yours (https://github.com/grpEHR/post-covid-events-ehrQL/blob/dataset-definition/analysis/metadates.R).
> 
> Locally, I can run each command individually (opensafely exec r:latest analysis/metadates.R --output output/study_dates.json, then opensafely exec ehrql:v1 generate-dataset analysis/dataset_definition.py --output output/dataset.arrow) and that works fine, but when I run all together (opensafely run run_all ) it breaks (see error message), must have something to do with the setup/yaml actions?
> 
> Any idea?

> try "opensafely run run_all -f" to force the command, see how it goes, i used this and run okay.

**Review by ZoeMZou — CHANGES_REQUESTED** [review]
Just minor changes are needed.

File: analysis/dataset_definition.py [hunk 1]
```diff
@@ -1,14 +1,672 @@
...
+  .where(vaccinations.date.is_on_or_before(baseline_date))
+  .sort_by(vaccinations.date)
+  .last_for_patient()
+  .date
+)
+
+## BMI, most recent value, within previous 2 years
+bmi_measurement = most_recent_bmi(
+    where=clinical_events.date.is_after(baseline_date - days(2 * 366)),
+    minimum_age_at_measurement=16,
```
Comment (ZoeMZou) [inline analysis/dataset_definition.py:520]: Do you really need to specify the minimum age of 16? you would exclude the participants aged <18 at baseline anyway in the later pipelines.

**Review by ZoeMZou — CHANGES_REQUESTED** [review]
The codes run successfully. To make the codes clearer, please see specific suggestions for each file.

**Review by ZoeMZou — CHANGES_REQUESTED** [review]
The codes run successfully locally. Just a few changes are needed to make them shorter and more efficient.

File: analysis/variable_helper_functions.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    )
+
+## DEATH
+def matching_death_before(codelist, baseline_date, where=True):
+    conditions = [
+        getattr(ons_deaths, column_name).is_in(codelist)
+        for column_name in (["underlying_cause_of_death"]+[f"cause_of_death_{i:02d}" for i in range(1, 16)])
+    ]
+    return(
+        ons_deaths.where()
```
[the remark this answers is not in the document]
Comment (alainamstutz) [reply analysis/variable_helper_functions.py]: done

File: analysis/codelists.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+# Gestational diabetes
+diabetes_gestational_ctv3_clinical = codelist_from_csv("codelists/user-hjforbes-gestational-diabetes.csv",column="code")
+# Type 1 diabetes secondary care
+diabetes_type1_icd10 = codelist_from_csv("codelists/opensafely-type-1-diabetes-secondary-care.csv",column="icd10_code")
+# Type 2 diabetes secondary care
+diabetes_type2_icd10 = codelist_from_csv("codelists/user-r_denholm-type-2-diabetes-secondary-care-bristol.csv",column="code")
+# Non-diagnostic diabetes codes
+diabetes_diagnostic_ctv3_clinical = codelist_from_csv("codelists/user-hjforbes-nondiagnostic-diabetes-codes.csv",column="code")
+# HbA1c
+hba1c_new_codes = codelist_from_csv("codelists/user-alainamstutz-hba1c-bristol.csv",column="code")
```
[the remark this answers is not in the document]
Comment (alainamstutz) [reply analysis/codelists.py:69]: thanks a lot - well spotted. I use it now.

**Pull request merged**